project_task_scheduling/wizards/scheduling_wizard.py

```diff
@@ -0,0 +1,397 @@
+                    break
+                start = max(state.tasks_dict[depend.id].end_datetime, start)
+
+            best_assigment = False
+            best_end_datetime = distant_datetime
+            for employee in task._get_employees():
+                interval, index = self._get_assignment(state, task, employee,
+                                                       start)
+                if interval and interval.end_datetime < best_end_datetime:
+                    interval.data.update(task=task, employee=employee)
+                    best_end_datetime = interval.end_datetime
+                    best_assigment = (interval, index)
+
+            if best_assigment:
+                interval, index = best_assigment
+
+                state.tasks_dict[task.id] = interval
+
+                employee = interval.data['employee']
+                state.employees_dict.setdefault(employee.id, [])
+                state.employees_dict[employee.id].insert(index, interval)
+
+        return state
+
+    def _generate_neighbor(self, state):
+        new_state = self._state_copy(state)
+        tasks_list = new_state.tasks_list
+        tasks_dict = new_state.tasks_dict
+        employees_dict = new_state.employees_dict
+
+        pos = random.randrange(len(tasks_list) - 1)
+        if tasks_list[pos] in tasks_list[pos + 1].dependency_task_ids:
+            return False
+
+        # swap
+        temp = tasks_list[pos]
+        tasks_list[pos] = tasks_list[pos + 1]
+        tasks_list[pos + 1] = temp
+
+        # delete task in pos and after pos
+        for task in tasks_list[pos:]:
+            if task.id in tasks_dict:
+                interval = tasks_dict[task.id]
+                del tasks_dict[task.id]
+                employees_dict[interval.data['employee'].id].remove(interval)
+
+        return self._greedy_distribution(state=new_state, init_index=pos)
+
+    def _get_init_state(self):
+        tasks_list = self._get_sorted_tasks()
+        tasks_dict, employees_dict = self._get_task_employee_dicts(tasks_list)
+        state = self._state_obj(tasks_list, tasks_dict, employees_dict, False)
+
+        return self._greedy_distribution(state)
+
+    def _print_summary(self, states_list, current_duration, max_duration,
+                       it_count, iterations):
+        _logger.info('********** SUMMARY **********')
+
+        improvements = len(states_list) - 1
+        _logger.info('improvements: %(imp)d' % {'imp': improvements})
+
+        duration = current_duration.total_seconds()
+        _logger.info('Duration: %(duration)f' % {'duration': duration})
+
+        if current_duration >= max_duration:
+            _logger.info('Stopped because Execution time >= maximum time')
+
+        it_count *= iterations
+        _logger.info('Number of iterations: %(it)d' % {'it': it_count})
+
+        _logger.info('*****************************')
+
+    def simulated_annealing(self, init_temp, final_temp, iterations,
```

Comment:
Did you consider to adopt some external library to implement this algorithm?
Did you also consider to adopt any other different algorithms for the allocation, eg.: any algorithm for a knapsack problem optimization? I see that the algorithm you adopted is a high resource consumption one.

Reply:
You may want to have a look at this:
https://developers.google.com/optimization/scheduling/employee_scheduling

Reply:
I have not seen any library that uses this algorithm.
If there are other algorithms for the assignment, but these are also of high levels of resources. It is a NP-hard problem and that algorithms are used that give meta-heuristic solutions, like this one (simulated annealing). There are variants of the knapsack problem that can be dynamic solutions, that is, they solve the problem in polynomial order, but that type of solutions can not be applied here.